src/script/interpreter.cpp

```diff
@@ -918,10 +928,56 @@ bool EvalScript(vector<vector<unsigned char> >& stack, const CScript& script, un
+                    int empty_sigs = 0;
                     for (int k = 0; k < nSigsCount; k++)
                     {
                         valtype& vchSig = stacktop(-isig-k);
                         scriptCode.FindAndDelete(CScript(vchSig));
+                        empty_sigs += !!vchSig.empty();
+                    }
+                    // If any signature is empty, then they all must be empty; or
+                    // if any signature is non-empty, then they all must be non-empty.
+                    if (require_valid_sigs && (empty_sigs > 0) && (empty_sigs != nSigsCount)) {
+                        return set_error(serror, SCRIPT_ERR_FAILED_SIGNATURE_CHECK);
+                    }
+
+                    // A bug in the original CHECKMULTISIG implementation caused an
+                    // extra item to be popped off the stack upon completion. When
+                    // SCRIPT_VERIFY_REQUIRE_VALID_SIGS is in effect, this otherwise
+                    // unused parameter is a bitfield indicating which pubkeys are NOT
+                    // used. With this hint we can avoid expensive signature
+                    // validation checks that might fail.
+                    std::bitset<20U> skipped_pubkeys(0);
+                    if (require_valid_sigs) {
+                        // There cannot be more than 20 pubkeys, so skipped_pubkeys
+                        // cannot be more than 20 unsigned bits, which fits fine
+                        // inside a 3-byte signed CScriptNum.
+                        int bits = CScriptNum(stacktop(-i), true, 3).getint();
+                        // Make sure that bits is within the numeric range of integers
+                        // corresponding to our bitfield length, so that we don't risk
+                        // any malleability or implementation / platform-defined behavior.
+                        if ((bits < 0) || (bits >= (1 << nKeysCount))) {
+                            return set_error(serror, SCRIPT_ERR_MULTISIG_SKIPPED_PUBKEYS);
+                        }
+                        // Convert to bitset via implicit constructor and assignment
+                        skipped_pubkeys = static_cast<unsigned>(bits);
+                        // For a k-of-n multisig, there musts be (n - k) unused
+                        // pubkeys. We require that the bits corresponding to these
+                        // pubkeys be set in the skipped_pubkeys field. Note in
+                        // particular that the corresponding bits must be set for
+                        // pubkeys in the final positions, if unused, even though
+                        // the signature verification loop below terminates early
+                        // in that situation.
+                        if (!( (!empty_sigs && (skipped_pubkeys.count() == (nKeysCount - nSigsCount))) ||
+                               ( empty_sigs && (skipped_pubkeys.count() == (nKeysCount))) ))
```

Comment:
Can we unwrap this? All the nots and parens are making me dizzy. We can even replace the two cases with
```C++
skipped_pubkeys.count() == nKeysCount - (empty_sigs ? 0 : nSigsCount)
```
and get
```C++
if (!(skipped_pubkeys.count() == nKeysCount - (empty_sigs ? 0 : nSigsCount)))
```
which we can invert to
```C++
if (skipped_pubkeys.count() != nKeysCount - (empty_sigs ? 0 : nSigsCount))
```

Reply:
Yes, that's much clearer. Thanks. I did something like this in the refactor.